### ISSUE TITLE
Ability to get current permissions level in Protoflux

### ISSUE THREAD

**Issue by ultrawidegamer** [Yellow-Dog-Man/Resonite-Issues#297]
### Is your feature request related to a problem? Please describe.

Currently there is no way to get a users current permission level via protoflux

### Describe the solution you'd like

i would like to have a protoflux node or a component that when you provide it a user it will provide their current permission level back. If this was provided we would be able to write code that is based solely on the permission level that users have.

### Describe alternatives you've considered

N/A

**Comment by JackTheFoxOtter** [comment]
Important to note here is that this node would return a string, and you cannot assume roles are always called the same (for example "Builder"), since those can be renamed / edited. What permissions a certain role provides is also dependant on the world, since that can be configured. A "UserRole" node does however make sense to me though, considering we also have the AssignRole node which can be used alongside VerifyJoinRequest.

**Comment by LeCloutPanda** [comment]
I would also like to second this as it would be lovely to have access to this information for example to only allowed items to spawn if spawned by an Admin etc.

**Comment by ShrikeAlvaron** [comment]
I know there's been resistance to this in the past as creating systems based off world permissions isn't considered secure due to builders being able to alter the code itself, but I really do think there's a demand for this and would like to see it added; Protoflux as it is should never be relied upon to create secure systems by itself without also taking into account the permissions of users in the world.

> Important to note here is that this node would return a string, and you cannot assume roles are always called the same (for example "Builder"), since those can be renamed / edited.

I don't believe they can?  It's possible I'm forgetting something, but I believe those names are set and ideally such a node would return an enum of them.


> I would also like to second this as it would be lovely to have access to this information for example to only allowed items to spawn if spawned by an Admin etc.

This is already possible, you can configure what's (dis)allowed for each permission level within the role slots of a world, one of which includes spawn permissions.

**Comment by art0007i** [comment]
> I don't believe they can? It's possible I'm forgetting something, but I believe those names are set and ideally such a node would return an enum of them.

it is in fact possible to change permission names. you can try it by opening a user inspector and clicking the up arrow next to role, there you can alter the permission name and it will even update in the session tab

**Comment by Nytra** [comment]
I think you can even remove roles and possibly add new ones?

**Comment by LeCloutPanda** [comment]
> I would also like to second this as it would be lovely to have access to this information for example to only allowed items to spawn if spawned by an Admin etc.

I was more so meaning the ability to just see what the name/type of role is because I would like to implement a check to see if an object is spawned by a admin/builder/guest etc and most people don't change the default role names, I know of one person who did change them so being able to read the role and just doing a simple equals check would be lovely to have

**Comment by ProbablePrime** [comment]
Thanks for making this, i did want to confirm that Shrike is correct. We've avoided doing this in the past due to the security implications.

If you plan to fill your world with "if builder" checks it can be problematic as those checks aren't secure etc.

We could look into something similar to the Join Verifier but that already has a lot of code on it to try to make that secure and even then it sometimes isn't. For example in: https://wiki.resonite.com/Beta_2023.10.4.165 I already had to fix it and that was a couple of days before launch :)

We want to do this, we just need to make sure we're doing it correctly.

**Comment by Nytra** [comment]
Nodes that can probe the permissions of a user may be better for this. For example a `CanUserEquipTool` node which takes a `ITool` reference and that will tell you if a particular user can equip something like the dev tool.

**Comment by JackTheFoxOtter** [comment]
@ProbablePrime I think the underlying issue is that people would use the role name to check for permissions if exposed because there isn't an alternative. So my suggestion would be creating nodes the return the individual atomic permissions for a give user instead, reflecting the settings in the permission components. Like, `AllowSpawningObjects` or `AllowSavingItems` from the `WorldPermissions` component (So nodes for "CanSpawnObjects" or "CanSaveItems" that take a user input). Once that's a given I don't see issues with having a `UserRoleName` node that outputs the string name of the role. That should be clear that it's only the NAME of the role, and doesn't have any performance impact.

**Comment by JackTheFoxOtter** [comment]
(Also apologies, I legit thought this issue was closed, which is why I didn't post my response here >.> I'm not fully awake yet)

**Comment by ProbablePrime** [comment]
The issue isn't the role name, or the permissions themselves. The issue is ProtoFlux or the DataModel allowing the result of any node that you or @Nytra described from being manipulated.

Imagine there's a very dangerous operation tied up to any permission related node controlled by a boolean from an output of that node.

I can just break the drive, and put in a true from another source and that dangerous operation runs.

-----

The solution here is #1103 which among other things will allow for safe areas within a hierarchy to run protoflux like you would like to use.

We unfortunately won't be able to implement any of the ideas in this issue as they'd all be superseded by #1103.

So unfortunately, for this particular item, it is dependent on #1103

**Comment by art0007i** [comment]
> Imagine there's a very dangerous operation tied up to any permission related node controlled by a boolean from an output of that node.

Well if I'm making a tool that needs to perform a very dangerous operation anyway it does not hurt to put a very rudimentary permission check, the alternative is to have no check on that dangerous operation at all.

I think if builder checks aren't a problem because non builders won't really be able to bypass those and builders have no reason to do it themselves.

**Comment by Frooxius** [comment]
It does hurt, because you're hiding very dangerous operation behind a rudimentary permission check - which can be easily bypassed, making the very dangerous operation exploitable, with no recourse or way to mitigate or fix it. That's the state we're trying to avoid.

Once we introduce something into the system, people will start building on top of it. And at that point it becomes very difficult to undo any design decisions without breaking tons of content.

Since this deals with permissions and security, we are not comfortable making those hasty design decisions, where we feel there will be a lot of damage and problems down the line, that will be hard to undo. 

Rather we want to make sure that whatever we put in and expose (notably through hard permission system) will be a good solid foundation that will provide certain security guarantees.
